cannon/mipsevm/state.go

```diff
@@ -70,6 +72,214 @@ func (s *State) EncodeWitness() StateWitness {
 	return out
 }
 
+func (s *State) Serialize(out io.Writer) error {
+	// Write the version byte to the buffer.
+	if err := binary.Write(out, binary.BigEndian, uint8(0)); err != nil {
+		return err
+	}
+
+	// Write the memory to the buffer.
+	serMemBuf := new(bytes.Buffer)
+	err := s.Memory.Serialize(serMemBuf)
+	if err != nil {
+		return err
+	}
+	serMemBytes := serMemBuf.Bytes()
+	serMemLen := uint32(len(serMemBytes))
```

Comment:
I'd missed this before, but this is really very unfortunate since we now need to hold a copy of the entire memory in RAM while serializing.  That was causing OOM failures originally until we started gzipping the pages individually to reduce the size.

I think we need to find a way to avoid this and actually stream the memory content.  The simplest thing would be for the memory serialisation to prefix its content with the number of pages to read and then read exactly those pages back in rather than reading until EOF.  Net result is the prefix becomes the number of pages (ie list length) rather than number of bytes in the memory, but now you can fully stream the content when reading and writing.

As this currently stands I think we'll run into memory issues again.